### ISSUE TITLE
sync sub-command won't remove merged branches

### ISSUE THREAD

**Issue by gmile** [github/hub#1304]
In the following sequence `hub sync` won't remove local branches, that were merged/removed on Github:

1. hub clone
2. hub fork
3. git checkout -b new-branch
4. (make & commit changes)
5. git push gmile new-branch
6. (merge the branch on github)
7. (remove the branch on github)
8. hub sync

After `hub sync` local branch is not removed. Only master is updated. Full output:

```
$ hub sync
remote: Counting objects: 26, done.
remote: Compressing objects: 100% (25/25), done.
remote: Total 26 (delta 1), reused 0 (delta 0), pack-reused 0
Updated branch master (was 4f2c4c9).
$
```

**Comment by mislav** [comment]
Thanks for reporting! This is right now expected behavior, but I definitely see how you were surprised by this because it seems logical that "new-branch" would get deleted.

`hub sync` right now only manages to track branches that have "upstream" configuration in git terms. This is achieved with `git push -u gmile new-branch`, or with `git checkout new-branch` as a shorthand for `git checkout -b new-branch --track gmile/new-branch`.

A branch that doesn't have upstream configuration is never deleted during `sync` because I couldn't figure out how to cleanup all merged branches without risking disrupting someone's flow by deleting branches that technically *appear* merged but that a person might not have wanted deleted just yet. Therefore, deletion is only restricted to branches that have upstream configuration, but whose remote branch (`gmile/new-branch`) is now missing (i.e. deleted).

**Comment by gmile** [comment]
@mislav thanks for explaining the behaviour!

Your examples suggest `hub sync` will remove local branches if they were set to track remote upstream during branch creation or push.

Should `hub sync` properly remove existing branches, created without --set-track-to, but modified to track an upstream branch afterwards? I just tried this and failed, somewhat. I took existing branch, and modified it to track a remote upstream branch of the same name. Here's a full sequence of steps I did:

```bash
$ (branch update-deps is not tracking an upstream branch)
$ git branch update-deps --set-upstream-to gmile/update-deps
$ (merge branch update-deps on github)
$ hub sync
$ (only see master branch updated)
```
This is what I see in `git branch -vv` after modifying the branch to track upstream:

```
$ git branch -vv
master      e807e13 [origin/master] Merge pull request #24 from gmile/update-deps
update-deps 2996ee3 [gmile/update-deps] Include timex_ecto package
$
```

**Comment by mislav** [comment]
Was `gmile/update-deps` merged and deleted on GitHub? These are the two prerequisites for `sync` to delete the local branch.

**Comment by jawshooah** [comment]
@mislav I'm having the same problem, but only for branches tracking remotes other than `origin`:

```bash
$ git init test-hub-sync
$ cd test-hub-sync
$ hub create
$ git commit --allow-empty -m 'Initial commit'
$ git push -u origin master
$ git checkout -b pr-branch
$ git commit --allow-empty -m 'PR branch commit'
$ git push -u origin pr-branch
$ hub pull-request
# merge PR and delete remote pr-branch
$ hub sync
remote: Counting objects: 1, done.
remote: Total 1 (delta 0), reused 0 (delta 0), pack-reused 0
Updated branch master (was fe196f7).
Deleted branch pr-branch (was 7a19ebc).
$ git checkout -b pr-branch-2
$ git commit --allow-empty -m 'PR branch commit 2'
$ git remote add other git@github.com:jawshooah/test-hub-sync.git
$ git push -u other pr-branch-2
$ hub pull-request
# merge PR and delete remote pr-branch-2
$ hub sync
remote: Counting objects: 1, done.
remote: Total 1 (delta 0), reused 0 (delta 0), pack-reused 0
Updated branch master (was 1b8ca48).
```

**Comment by mislav** [comment]
@jawshooah `sync` currently only fetches from main remote, and ignores other remotes. Should `sync` try to fetch from all configured remotes and affect branches that are tracking either of those remotes?

**Comment by jawshooah** [comment]
> Should sync try to fetch from all configured remotes and affect branches that are tracking either of those remotes?

I think so, yes. I almost always open PRs from my own fork rather than from a branch on the source repo, so it would be very nice if `sync` could update branches that track those as well.

**Comment by jawshooah** [comment]
Related: how does `sync` determine which remote is the "main" remote? Is it just the remote tracked by local `master`?

**Comment by mislav** [comment]
It's the first remote named either "upstream", "github", or "origin", in that order.

**Comment by gmile** [comment]
> @mislav Was gmile/update-deps merged and deleted on GitHub

Yes, it was. It's weird that `hub sync` didn't delete it locally.

> @mislav: Should sync try to fetch from all configured remotes and affect branches that are tracking either of those remotes?

That would be good, but I think there's a simpler middle ground here.

I think it would be good if `hub sync` follows the "happy path" of `hub`: if you don't specify any args, all commands work with the same assumptions e.g. the "same way". As such, I would expect `hub pull-request` and `hub sync` to work in conjunction. What I'd love to see is this:

1. `hub pull-request` (without arguments) by default creates a fork under my own account (instead of under organization account, e.g. `origin`) – **we have this**,
2. `hub sync`(without arguments) by default cleans up branches under (but not limited to) my own account's fork, made using hub pull-request – **I'd love to have this**.

**Comment by mislav** [comment]
> It's weird that `hub sync` didn't delete it locally.

After running `hub sync`, does `gmile/update-deps` still exist in `git branch -r` list? If so, that might be the cause of the problem, because `git fetch --prune` (that sync does as a first operation) is supposed to first remove all remote tracking branches that have been deleted in the GitHub repository.

> 2\. `hub sync`(without arguments) by default cleans up branches under (but not limited to) my own account's fork, made using hub pull-request – **I'd love to have this**.

That's a good suggestion! Right now it doesn't clean up those branches that you submitted through your fork but that got merged in the meantime. It definitely should.

**Comment by mislav** [comment]
Ah, I just realized that you might have multiple git remotes and that "gmile" is the remote for your fork, while `hub sync` will usually just run operations against the "origin" or "upstream" remote. That explains why the `gmile/*` branches didn't get cleaned up. I definitely need to think how to implement this feature. Or, you could take a stab at submitting a PR! 😉 But in any case, thanks for the productive discussion.

**Comment by wesbland** [comment]
If you're concerned about accidentally deleting branches that the user may not have wanted to delete, you could add a `--prune` option to `hub sync`. By default, it wouldn't delete the local branch unless you explicitly ask it to do so.

**Comment by jawshooah** [comment]
You could also keep the current behavior of only fetching from the main remote as a default, and fetch from all remotes when given a flag like `--all`.
